Snapshot-Based Node Split – Foundation and Core Implementation

codeforall opened the pull request:
**DESCRIPTION:**
This pull request introduces the foundation and core logic for the snapshot-based node split feature in Citus. This feature enables promoting a streaming replica (referred to as a clone in this feature and UI) to a primary node and rebalancing shards between the original and the newly promoted node without requiring a full data copy.

This significantly reduces rebalance times for scale-out operations where the new node already contains a full copy of the data via streaming replication.

### Key Highlights
**1. Replica (Clone) Registration & Management Infrastructure**

Introduces a new set of UDFs to register and manage clone nodes:

- citus_add_clone_node()
- citus_add_clone_node_with_nodeid()
- citus_remove_clone_node()
- citus_remove_clone_node_with_nodeid()

These functions allow administrators to register a streaming replica of an existing worker node as a clone, making it eligible for later promotion via snapshot-based split.

**2. Snapshot-Based Node Split (Core Implementation)**
New core UDF: 

- citus_promote_clone_and_rebalance()

This function implements the full workflow to promote a clone and rebalance shards between the old and new primaries. Steps include:

1. Ensuring Exclusivity – Blocks any concurrent placement-changing operations.
2. Blocking Writes – Temporarily blocks writes on the primary to ensure consistency.
3. Replica Catch-up – Waits for the replica to be fully in sync.
4. Promotion – Promotes the replica to a primary using pg_promote.
5. Metadata Update – Updates metadata to reflect the newly promoted primary node.
6. Shard Rebalancing – Redistributes shards between the old and new primary nodes.


**3. Split Plan Preview**
A new helper UDF get_snapshot_based_node_split_plan() provides a preview of the shard distribution post-split, without executing the promotion.

**Example:**

```
reb 63796> select * from pg_catalog.get_snapshot_based_node_split_plan('127.0.0.1',5433,'127.0.0.1',5453);
  table_name  | shardid | shard_size | placement_node 
--------------+---------+------------+----------------
 companies    |  102008 |          0 | Primary Node
 campaigns    |  102010 |          0 | Primary Node
 ads          |  102012 |          0 | Primary Node
 mscompanies  |  102014 |          0 | Primary Node
 mscampaigns  |  102016 |          0 | Primary Node
 msads        |  102018 |          0 | Primary Node
 mscompanies2 |  102020 |          0 | Primary Node
 mscampaigns2 |  102022 |          0 | Primary Node
 msads2       |  102024 |          0 | Primary Node
 companies    |  102009 |          0 | Clone Node
 campaigns    |  102011 |          0 | Clone Node
 ads          |  102013 |          0 | Clone Node
 mscompanies  |  102015 |          0 | Clone Node
 mscampaigns  |  102017 |          0 | Clone Node
 msads        |  102019 |          0 | Clone Node
 mscompanies2 |  102021 |          0 | Clone Node
 mscampaigns2 |  102023 |          0 | Clone Node
 msads2       |  102025 |          0 | Clone Node
(18 rows)

```
**4 Test Infrastructure Enhancements**

- Added a new test case scheduler for snapshot-based split scenarios.
- Enhanced pg_regress_multi.pl to support creating node backups with slightly modified options to simulate real-world backup-based clone creation.

### 5. Usage Guide
The snapshot-based node split can be performed using the following workflow:

**- Take a Backup of the Worker Node**
Run pg_basebackup (or an equivalent tool) against the existing worker node to create a physical backup.

`pg_basebackup -h <primary_worker_host> -p <port> -D /path/to/replica/data --write-recovery-conf
`

**- Start the Replica Node**
Start PostgreSQL on the replica using the backup data directory, ensuring it is configured as a streaming replica of the original worker node.

**- Register the Backup Node as a Clone**
Mark the registered replica as a clone of its original worker node:

`SELECT * FROM citus_add_clone_node('<clone_host>', <clone_port>, '<primary_host>', <primary_port>);
`

**- Promote and Rebalance the Clone**
Promote the clone to a primary and rebalance shards between it and the original worker:

`SELECT * FROM citus_promote_clone_and_rebalance('clone_node_id');
`

**- Drop Any Replication Slots from the Original Worker**
After promotion, clean up any unused replication slots from the original worker:

`SELECT pg_drop_replication_slot('<slot_name>');
`


### Impact & Benefits

- Faster Scaling: Eliminates the need for full data copy when adding a new node from an existing replica.
- Reduced Downtime: Shorter rebalance time means less application disruption.
- Operational Simplicity: UDF-driven workflow for easy orchestration.

eaydingol commented on src/backend/distributed/operations/node_promotion.c:
```diff
@@ -0,0 +1,449 @@
+							primaryNode->workerName, primaryNode->workerPort)));
+
+	bool caughtUp = false;
+	const int catchUpTimeoutSeconds = 300; /* 5 minutes, TODO: Make GUC */
```
It would be nice to have this as a GUC or a udf parameter

codeforall replied:
I have added a new parameter to udf with 300 seconds as default value.

eaydingol commented on src/backend/distributed/operations/shard_rebalancer.c:
```diff
@@ -81,8 +81,29 @@ typedef struct RebalanceOptions
+	List *primaryShardIdList;
+
+	/*
+	 * replicaShardPlacementList contains the placements that should stay on
```
maybe  add "(clone)" to comment to emphasize the connection

codeforall replied:
I’ve updated this and several other variables to reflect the new naming convention, and also added a few additional NOTICE messages to make the purpose and flow of operations clearer.

eaydingol commented on src/backend/distributed/operations/shard_transfer.c:
```diff
@@ -574,6 +574,95 @@ TransferShards(int64 shardId, char *sourceNodeName,
+													   replicaNode->workerName,
+													   replicaNode->workerPort);
+
+		/* Now drop all shards from primary that need to be on the clone node */
```
Can you also write a comment on at which point the shards that are on primary ( or on replica) are cleaned up from the replica (or from primary)

codeforall replied:
I've added the appropriate NOTICE message. Kindly check if it is more clear now?

eaydingol commented:
The overall idea is great and has strong potential to scale citus cluster size more efficiently. 
Given the critical nature of this functionality, we need additional test coverage. Specifically, edge case scenarios such as an empty source node or a source node containing only reference tables would be good.

It is useful to add unit tests for each new UDF. 

The code already includes logic to handle failures at various points, which is great. We also need failure scenarios in the test suite, possibly using mitmproxy or simulating shard locks.  In particular, the main flow for promote-clone and rebalance includes non-transactional statements. It would be helpful to have failure scenarios that test these paths and confirm that the cluster remains in a recoverable state when a failure occurs.

codeforall commented on src/backend/distributed/operations/shard_rebalancer.c:
```diff
@@ -3547,6 +3579,357 @@ EnsureShardCostUDF(Oid functionOid)
+		.workerNode = primaryNode /* indicate Primary node as a source node */
+	};
+
+	SplitPrimaryReplicaShards *splitShards = GetPrimaryReplicaSplitRebalanceSteps(&options
```
GetPrimaryReplicaSplitRebalanceSteps always returns a palloc’d structure, so it can never be NULL. Infect, the check in get_snapshot_based_node_split_plan was unnecessary, and I’ve removed it.

codeforall commented:
> The overall idea is great and has strong potential to scale citus cluster size more efficiently. Given the critical nature of this functionality, we need additional test coverage. Specifically, edge case scenarios such as an empty source node or a source node containing only reference tables would be good.
> 
> It is useful to add unit tests for each new UDF.
> 
> The code already includes logic to handle failures at various points, which is great. We also need failure scenarios in the test suite, possibly using mitmproxy or simulating shard locks. In particular, the main flow for promote-clone and rebalance includes non-transactional statements. It would be helpful to have failure scenarios that test these paths and confirm that the cluster remains in a recoverable state when a failure occurs.

Thanks a lot for the thoughtful review!

I’ve made a few follow-up changes based on your feedback:

Code improvements: Reorganized some parts and added extra validation around verifying a clone node before it can be registered or promoted.

Workflow clarity: Added additional NOTICE messages to make the flow of operations more visible when running the UDFs.

Test coverage: Extended the test suite to cover failure and error scenarios. These include cases where promotion fails or where a transaction is rolled back after promote-clone. The tests also verify that the cluster remains in a consistent and recoverable state in those situations.

I agree with your suggestion about using mitmproxy (or similar techniques) for simulating more complex failure scenarios. That’s a great idea, and I’d like to propose tackling it in a separate PR so we can keep this one focused.

codeforall commented on .github/workflows/build_and_test.yml:
```diff
@@ -489,7 +490,7 @@ jobs:
         skipped_tests=""
```
Done!
